File: src/main/java/org/jcodec/containers/mxf/model/GenericTrack.java
package org.jcodec.containers.mxf.model;
import java.util.Iterator;

import org.jcodec.common.logging.Logger;

import java.nio.ByteBuffer;
import java.util.Map;
import java.util.Map.Entry;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class GenericTrack extends MXFInterchangeObject {
    private int trackId;
    private String name;
    private UL sequenceRef;
    private int trackNumber;

    public GenericTrack(UL ul) {
        super(ul);
    }

    @Override
    protected void read(Map<Integer, ByteBuffer> tags) {
        for (Iterator<Entry<Integer, ByteBuffer>> it = tags.entrySet().iterator(); it.hasNext();) {
            Entry<Integer, ByteBuffer> entry = it.next();

            ByteBuffer _bb = entry.getValue();
            switch (entry.getKey()) {
            case 0x4801:
                trackId = _bb.getInt();
                break;
            case 0x4802:
                name = readUtf16String(_bb);
                break;
            case 0x4803:
                sequenceRef = UL.read(_bb);
                break;
            case 0x4804:
                trackNumber = _bb.getInt();
                break;
            default:
                Logger.warn(String.format("Unknown tag [ " + ul + "]: %04x", entry.getKey()));
                continue;
            }
            it.remove();
        }
    }

    public int getTrackId() {
        return trackId;
    }

    public String getName() {
        return name;
    }

    public UL getSequenceRef() {
        return sequenceRef;
    }

    public int getTrackNumber() {
        return trackNumber;
    }
}

File: src/main/java/org/jcodec/containers/mxf/model/GenericTrack.kt
package org.jcodec.containers.mxf.model

import org.jcodec.common.logging.Logger
import org.jcodec.containers.mxf.model.UL.Companion.read
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
open class GenericTrack(ul: UL?) : MXFInterchangeObject(ul) {
    var trackId = 0
        private set
    var name: String? = null
        private set
    var sequenceRef: UL? = null
        private set
    var trackNumber = 0
        private set

    override fun read(tags: MutableMap<Int, ByteBuffer>) {
        val it: MutableIterator<Map.Entry<Int, ByteBuffer>> = tags.entries.iterator()
        loop@ while (it.hasNext()) {
            val entry = it.next()
            val _bb = entry.value
            when (entry.key) {
                0x4801 -> trackId = _bb.int
                0x4802 -> name = readUtf16String(_bb)
                0x4803 -> sequenceRef = read(_bb)
                0x4804 -> trackNumber = _bb.int
                else -> {
                    Logger.warn(String.format("Unknown tag [ $ul]: %04x", entry.key))
                    continue@loop
                }
            }
            it.remove()
        }
    }

}